lib/icingadb/icingadbchecktask.cpp

```diff
@@ -0,0 +1,386 @@
+					"0-0", "0-0", "0-0", "0-0", "0-0", "0-0"
+				}
+			},
+			RedisConnection::QueryPriority::Heartbeat
+		));
+
+		xReadHeartbeat = std::move(replies.at(0));
+		xReadStats = std::move(replies.at(1));
+		xReadHistory = std::move(replies.at(2));
+	} catch (const std::exception& ex) {
+		ReportIcingadbCheck(checkable, commandObj, cr, String("Could not read XREAD responses from Redis: ") + ex.what(), ServiceCritical);
+		return;
+	}
+
+	if (!xReadHeartbeat) {
+		ReportIcingadbCheck(
+			checkable, commandObj, cr,
+			"The Icinga DB daemon seems to have never run. (Missing heartbeat)",
+			ServiceCritical
+		);
+
+		return;
+	}
+
+	Array::Ptr heartbeatMessage = Array::Ptr(Array::Ptr(xReadHeartbeat->Get(0))->Get(1))->Get(0);
+	auto heartbeatTime (GetXMessageTs(heartbeatMessage));
+	std::map<String, String> heartbeatData;
+
+	IcingaDB::AddKvsToMap(heartbeatMessage->Get(1), heartbeatData);
+
+	String version = heartbeatData.at("meta:version");
+	Dictionary::Ptr goMetricsByCumulativity (JsonDecode(heartbeatData.at("go:metrics")));
+	String dbErr (heartbeatData.at("db:err"));
+	auto ourHeartbeatTs (Convert::ToLong(heartbeatData.at("heartbeat:last-ts")) / 1000.0);
+	bool weResponsible = Convert::ToLong(heartbeatData.at("ha:we-responsible"));
+	auto weResponsibleTs (Convert::ToLong(heartbeatData.at("ha:we-responsible-ts")) / 1000.0);
+	bool otherResponsible = Convert::ToLong(heartbeatData.at("ha:other-responsible"));
+
+	auto now (Utility::GetTime());
+	auto downFor (now - heartbeatTime);
+	auto idleFor ((weResponsible ? -1 : 1) * (now - weResponsibleTs));
+	double historyBacklog = 0;
+
+	if (xReadHistory) {
+		double minTs = 0;
+		ObjectLock lock (xReadHistory);
+
+		for (Array::Ptr stream : xReadHistory) {
+			auto ts (GetXMessageTs(Array::Ptr(stream->Get(1))->Get(0)));
+
+			if (minTs == 0 || ts < minTs) {
+				minTs = ts;
+			}
+		}
+
+		if (minTs > 0) {
+			historyBacklog = now - minTs;
+		}
+	}
+
+	Array::Ptr perfdata = new Array();
+	std::map<String, RingBuffer> statsPerOp;
+
+	for (auto metric : icingadbKnownStats) {
+		statsPerOp.emplace(metric, 15 * 60);
+	}
+
+	if (xReadStats) {
+		Array::Ptr messages = Array::Ptr(xReadStats->Get(0))->Get(1);
+		ObjectLock lock (messages);
+
+		for (Array::Ptr message : messages) {
+			auto ts (GetXMessageTs(message));
+			std::map<String, String> opsPerSec;
+
+			IcingaDB::AddKvsToMap(message->Get(1), opsPerSec);
+
+			for (auto& kv : opsPerSec) {
+				auto buf (statsPerOp.find(kv.first));
+
+				if (buf == statsPerOp.end()) {
+					buf = statsPerOp.emplace(kv.first, 15 * 60).first;
+				}
+
+				buf->second.InsertValue(ts, Convert::ToLong(kv.second));
+			}
+		}
+	}
+
+	ServiceState state = ServiceOK;
+	std::ostringstream msgbuf;
+	double qps = redis->GetQueryCount(60) / 60.0;
+	double pendingQueries = redis->GetPendingQueryCount();
+
+	auto checkLower ([&state, &msgbuf](
+		double value, const Value& warning, const String& missingWarning,
+		const Value& critical, const String& missingCritical
+	) {
```

Comment:
Could be updated to take two `const Threshold&` as parameters instead of the four individual values. Or even better: just a single `const Thresholds&` (but this needs some more unification in threshold handling than I proposed in previous comments, but doing it would probably be a good thing).